blockstack/stacks-blockchain
# Serve pollable prometheus endpoint

This PR is addressing https://github.com/blockstack/stacks-blockchain/issues/1432, and spawning an endpoint, exposing metrics that can be scraped by prometheus.

### How to test:
Easiest way to start at this point, is by spawning a testnet, with nodes compiled w/ feature flag `monitoring`, ie --

Start bitcoind:
```
bitcoind
```
Start a bitcoin controller:
```
cd testnet/bitcoin-neon-controller
DYNAMIC_GENESIS_TIMESTAMP=1 cargo run local-leader.toml.default
```

Assuming prometheus is installed on your machine, start it with the sample config:
```
prometheus --config.file="testnet/stacks-node/conf/prometheus.yml" 
```

Start a local leader node:
```
cd testnet/stacks-node
cargo run  --features "monitoring" start --config=./conf/local-leader-conf.toml
```

Start a local follower node:
```
cd testnet/stacks-node
cargo run  --features "monitoring" start --config=./conf/local-follower-conf.toml
```

When visiting `http://127.0.0.1:4000` and `http://127.0.0.1:5000` you can see the metrics exposed (respectively miner and follower), that can be scraped by prometheus.

You can also visit `http://0.0.0.0:9090/` and query things like 
```
sum(stacks_node_btc_blocks_received_total)
```

I wanter to start with a set of simple metrics so we can settle on the way prometheus is being integrated, we can get more fancy with more metrics, histograms and gauges in a near future.

Once debated, adjusted, accepted, merged and deployed, I'll probably be following-up with another PR, implementing some more adjustments required by DevOps team.
